charm_temp_errors.md

```diff
@@ -0,0 +1,114 @@
+We're pretty sure that when we do get this failure, it's a temporary problem - perhaps there's high load, or some temporary networking issues. One issue with using `defer()` is that we don't know, or have any control over, when the event handler will be retried. `defer()` also restarts the entire handler, so any work we've already done will get repeated, which is wasteful. Let's adjust the code so that we give Pebble a few chances:
+
+```python
+for attempt in range(MAX_PEBBLE_RETRIES):
+    try:
+        container.replan()
+    except ops.pebble.ConnectionError:
+        time.sleep(PEBBLE_RETRY_DELAY ** attempt)
+    else:
+        break 
+else:
+    event.defer()
+    return
+```
+
+[tenacity](https://tenacity.readthedocs.io/en/latest/) is a really nice library providing decorators and context managers that handle different backoff mechanisms, attempt limits, and so forth - but if you want to keep your dependencies to a minimum, it's simple enough to have something like the above as reusable code within your charm.
+
+At this point, if Pebble is responding, then we'll get our `replan()` executed in essentially the same time as if we just called it (certainly faster than if we guarded with `can_connect()`), and if there's an issue with Pebble, we'll retry until we're confident that the problem isn't going to resolve itself within a timeframe we're comfortable with (perhaps a couple of minutes - remembering that the Charm won't handle other events while it's sleeping between retries). If Pebble is failing longer than that, do we really want to defer? That would be appropriate if we believe that the issue will resolve itself, but if Pebble has been unreachable this long, it's likely that someone needs to investigate what's happening - ie. manual intervention is required, and a Juju error state is appropriate in that case. Let's adjust the code again:
+
+```python
+for attempt in range(MAX_PEBBLE_RETRIES):
+    try:
+        container.replan()
+    except ops.pebble.ConnectionError:
+        time.sleep(PEBBLE_RETRY_DELAY ** attempt)
+    else:
+        break 
+else:
+    logger.error("Unable to reach Pebble after %.1f seconds.", too_lazy_to_do_the_maths)
+    raise SystemExit(1)
+```
+
+The only time we aren't expecting Pebble to be available is before we get the container's `PebbleReadyEvent`. If our code is running in an event handler that's potentially before then (see [a charm's lifecycle](https://juju.is/docs/sdk/charm-lifecycle)), we'd want to have the work be handled by a later event (or defer). If the code is running in a handler than could occur in the "setup" as well as "operation" or "teardown" phases, then we might prefer to always avoid the error state.
+
+Let's put it all together into a simple, reusable, method:
+
+.. TODO: I don't love how this is. A decorator or context manager is likely best. Maybe the handling of defer, return etc is too much? A class with a __getattr__ could make the call look much more like the normal version.
```

Comment:
Yeah, I'm not sure I love the `tolerant_pebble` proposal either. It seems to do too much in one tool, really two separate things: the retrying of the Pebble call, and the handling of an error. Might be worth a quick voice discussion to brainstorm whether we want this or how it should look.